### PR TITLE
fix: rework ops.main type hints to allow different flavours (protocol)

### PR DESCRIPTION
**Pull request by dimaqq**
- [x] our project tests
- [ ] super-tox
- [ ] deprecation docs

### CODE REVIEW

**Comment by dimaqq** [comment]
Doc build fails with `/code/operator/ops/main.py:docstring of ops.main:1: WARNING: duplicate object description of ops.main, other instance in index, use :no-index: for one of them`

I think it's because there's ops/main and ops/_main both with a docstring.

Should we hide the legacy bit doc string? Or is there a way to rename it for docs only?

File: test/test_main_type_hint.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Callable, Type
+
+import ops
+
+
+def type_test_dummy(_arg: Callable[[Type[ops.CharmBase], bool], None]):
+    """
+    Helper to verify that
```
Comment (tonyandrewmeyer) [inline test/test_main_type_hint.py:21]: More words expected?

File: test/test_main_invocation.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    (tmp_path / 'dispatch').absolute().touch(mode=0o755)
+
+    yield
+
+    os.environ.pop('OPERATOR_DISPATCH', None)
+
+
+class IdleCharm(ops.CharmBase):
+    def __init__(self, framework: ops.Framework):
+        super().__init__(framework)
```
Comment (tonyandrewmeyer) [inline test/test_main_invocation.py:44]: Why not just use `CharmBase`?

File: ops/main.py [hunk 3]
```diff
@@ -11,529 +11,18 @@
...
 
 import ops.charm
-import ops.framework
-import ops.model
-import ops.storage
-from ops.charm import CharmMeta
-from ops.jujucontext import _JujuContext
-from ops.log import setup_root_logging
-
-CHARM_STATE_FILE = '.unit-state.db'
```
Comment (tonyandrewmeyer) [inline ops/main.py:39]: Probably not importable any more?

File: ops/main.py [hunk 4]
```diff
@@ -11,529 +11,18 @@
...
-from typing import Any, Dict, List, Optional, Tuple, Type, Union, cast
+from typing import Optional, Type
 
 import ops.charm
-import ops.framework
-import ops.model
-import ops.storage
-from ops.charm import CharmMeta
-from ops.jujucontext import _JujuContext
-from ops.log import setup_root_logging
```
Comment (tonyandrewmeyer) [inline ops/main.py:37]: Probably should still be able to import CharmMeta and setup_root_logging.

File: ops/_main.py [hunk 5]
```diff
@@ -0,0 +1,555 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Main entry point to the framework.
+
+Note that this module is callable, and calls the :func:`ops.main.main` function.
```
Comment (tonyandrewmeyer) [inline ops/_main.py:17]: Outdated?

**Review by tonyandrewmeyer — COMMENTED** [review]
More-or-less the same comments as on the callable class version.

I'd probably want to re-review if this ends up being chosen.

**Comment by dimaqq** [comment]
Closing in favour of #1345

**Pull request closed without merging**